setup.rst

```diff
@@ -385,10 +385,11 @@ for the header and library files to your ``configure`` command.  For example,
 
 with **Homebrew**::
 
-    $ brew install openssl xz gdbm
+    $ brew install openssl xz gdbm tcl-tk
 
 and ``configure`` Python versions >= 3.7::
 
+    PKG_CONFIG_PATH="$(brew --prefix tcl-tk)/lib/pkgconfig"
```

Comment:
`configure` as of **Python 3.11** uses pkg-config. For pre 3.11 `configure`, you'll need to use something like this, IIRC:

```console
$ ./configure --with-tcltk-includes=$(brew --prefix tcl-tk)/include --with-tcltk-libs=$(brew --prefix tcl-tk)/lib
```

Please test the above before adding it.

See also discussion on [Discourse as of March 2022](https://discuss.python.org/t/simplifying-tcl-tk-detection/14066?u=erlendaasland)

Reply:
@erlend-aasland thanks very much.
Two things: I am unclear about what to test and how to test it. 

----
This is what I understood:
- For Python versions >= 3.11

I keep it as it is originally suggested in this PR:
```` 
PKG_CONFIG_PATH="$(brew --prefix tcl-tk)/lib/pkgconfig" 
./configure --with-pydebug --with-openssl=$(brew --prefix openssl) 
````
- For 3.11 > Python versions >= 3.7:

The line you suggested in the above comment (to be tested):
````
$ ./configure --with-tcltk-includes=$(brew --prefix tcl-tk)/include --with-tcltk-libs=$(brew --prefix tcl-tk)/lib
````
followed by the original options provided in the documentation:
````
--with-pydebug --with-openssl=$(brew --prefix openssl)
````

- For Python versions < 3.7

Documentation currently says:

````
$ CPPFLAGS="-I$(brew --prefix openssl)/include" \
LDFLAGS="-L$(brew --prefix openssl)/lib" \
./configure --with-pydebug
````
It would need to be updated as follow:
````
$ CPPFLAGS="-I$(brew --prefix openssl)/include -I$(brew --prefix tcl-tk)/include" \
LDFLAGS="-L$(brew --prefix openssl)/lib -L$(brew --prefix  tcl-tk)/lib" \
./configure --with-pydebug
````
The above is my understanding, let me know if this is correct.

----

> Please test the above before adding it.

If I test using 3.10.4, would that be enough for the use case `3.11 > Python versions >= 3.7`? I am currently on 3.10.4. 
Should I test for < 3.7 as well? I am quite new to Python, so I would uninstall my current version of Python and install 3.6 to test it as I am not confident running multiple versions at this time.
Should I test for 3.11?

These are the steps I would follow to test:
`make clean`, followed by the `./configure` line to be tested, and `make -s -j2`.

Reply:
Note, you don't need to uninstall or install a specific Python version to test these things. You only need to clone the git repo, check out the version you wish to test, then `./configure` and either manually inspect the Makefile, or just build (run `make`).

Reply:
To make it more readable and easier to copy/paste, the command above can be rewritten as:
```sh
PKG_CONFIG_PATH="$(brew --prefix tcl-tk)/lib/pkgconfig" \
./configure --with-tcltk-libs="$(pkg-config --libs tcl tk)" \
            --with-tcltk-includes="$(pkg-config --cflags tcl tk)"
```
Since this is a single command, the `export` shouldn't be necessary anymore.

Reply:
That don't work on my system; that's why I added the explicit `export`:

```console
$ PKG_CONFIG_PATH="$(brew --prefix tcl-tk)/lib/pkgconfig" \
    ./configure --with-tcltk-libs="$(pkg-config --libs tcl tk)" \
                --with-tcltk-includes="$(pkg-config --cflags tcl tk)"
Package tcl was not found in the pkg-config search path.
Perhaps you should add the directory containing `tcl.pc'
to the PKG_CONFIG_PATH environment variable
No package 'tcl' found
Package tk was not found in the pkg-config search path.
Perhaps you should add the directory containing `tk.pc'
to the PKG_CONFIG_PATH environment variable
No package 'tk' found
Package tcl was not found in the pkg-config search path.
Perhaps you should add the directory containing `tcl.pc'
to the PKG_CONFIG_PATH environment variable
No package 'tcl' found
Package tk was not found in the pkg-config search path.
Perhaps you should add the directory containing `tk.pc'
to the PKG_CONFIG_PATH environment variable
No package 'tk' found

```

Reply:
Thanks for verifying that!
I think that's because of the `$(...)`s in the `configure` args -- even though the `PKG_CONFIG_PATH` should be available while `configure` is executed, it's not yet available there.

This similar example seems to prove my point:
```sh
$ FOO=123 python3 -c 'import os;print(os.environ["FOO"])'
123
$ FOO=123 echo $(python3 -c 'import os;print(os.environ["FOO"])')
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.10/os.py", line 679, in __getitem__
    raise KeyError(key) from None
KeyError: 'FOO'
```

@izumiberat, you can ignore my comment about removing the `export`, but the next command could still be formatted in two lines:
```sh
$ export PKG_CONFIG_PATH="$(brew --prefix tcl-tk)/lib/pkgconfig"
$ ./configure --with-tcltk-libs="$(pkg-config --libs tcl tk)" \
              --with-tcltk-includes="$(pkg-config --cflags tcl tk)"
```

Reply:
I tested using `checkout` as suggested by @erlend-aasland for Python 3.4, 3.6, and 3.10 (as well as 3.8 just for fun).

Result : _tkinter gets built (this was the purpose of this PR), but I get the following warning for 3.10:

> configure: WARNING: unrecognized options: --with-tcltk-libs, --with-tcltk-includes

@erlend-aasland when closing issue https://github.com/python/cpython/issues/93659 I had tested that _tkinter got built but I hadn't provided the additional options `--with-tcltk-libs` and  `--with-tcltk-includes` in `configure`. I don't know and haven't researched yet the purpose of defining those options in `configure`. Can we omit them?

----

I copy/paste below part of the content displayed when `configure` and `make` for the different versions I tested.

**Summary**: There are a number of other issues that come up (out of the scope of this PR that I copy/paste below for reference) but it did remove the following message (which is the purpose of this PR):

```
Failed to build these modules:
_tkinter 
```    
For Python 3.4
----
> 9 warnings generated.
> ld: warning: could not create compact unwind for _ffi_call_unix64: does not use RBP or RSP based frame
> 
> Python build finished successfully!
> The necessary bits to build these optional modules were not found:
> _hashlib              _sqlite3              nis                
> ossaudiodev           spwd                  zlib               
> To find the necessary bits, look in setup.py in detect_modules() for the module's name.
> 
> The following modules found by detect_modules() in setup.py, have been
> built by the Makefile instead, as configured by the Setup files:
> atexit                pwd                   time   

For Python 3.6
----
same as above but with in addition:

> configure: WARNING: unrecognized options: --with-openssl

For Python 3.8
----
same as above but with:
_hashlib replaced by -ssl in:
> The necessary bits to build these optional modules were not found:
> _hashlib              _sqlite3              nis                
> ossaudiodev           spwd                  zlib    
 
For Python 3.10
----
`configure: WARNING: unrecognized options: --with-tcltk-libs, --with-tcltk-includes`

This is the full output when `make`:

> Note: Deepfreeze may have added some global objects,
>       so run 'make regen-global-objects' if necessary.
> /Library/Developer/CommandLineTools/usr/bin/ranlib: file: libpython3.12d.a(dynamic_annotations.o) has no symbols
> /Library/Developer/CommandLineTools/usr/bin/ranlib: file: libpython3.12d.a(pymath.o) has no symbols
>  CC='gcc' LDSHARED='gcc -bundle -undefined dynamic_lookup    ' OPT='-g -Og -Wall' 	./python.exe -E ./setup.py -q build
> ld: warning: directory not found for option '-L/usr/lib/termcap'
> renaming build/scripts-3.12/pydoc3 to build/scripts-3.12/pydoc3.12
> renaming build/scripts-3.12/idle3 to build/scripts-3.12/idle3.12
> renaming build/scripts-3.12/2to3 to build/scripts-3.12/2to3-3.12

Reply:
It may be that setup.py automatically detects and uses the system provided Tcl/Tk libraries on macOS. I'm not on my laptop now, so I cannot verify it (yet).

Reply:
@erlend-aasland when running `python3 -m tkinter`, it displays that I am using Tcl/Tk version 8.6. Looking at setup.py, it looks like, it is indeed installed by default if it doesn't find the options:

>         # Assume we haven't found any of the libraries or include files
>         # The versions with dots are used on Unix, and the versions without
>         # dots on Windows, for detection by cygwin.
>         tcllib = tklib = tcl_includes = tk_includes = None
>         for version in ['8.6', '86', '8.5', '85', '8.4', '84', '8.3', '83',
>                         '8.2', '82', '8.1', '81', '8.0', '80']:
>             tklib = self.compiler.find_library_file(lib_dirs,
>                                                         'tk' + version)
>             tcllib = self.compiler.find_library_file(lib_dirs,
>                                                          'tcl' + version)
>             if tklib and tcllib:
>                 # Exit the loop when we've found the Tcl/Tk libraries
>                 break

Not sure why specifying the options as you provided them lead to the warning:

> configure: WARNING: unrecognized options: --with-tcltk-libs, --with-tcltk-includes

Reply:
> Not sure why specifying the options as you provided them lead to the warning:

Something is wrong on your end. It may be wrong checkout (maybe your were on `main` instead of 3.10?).

```console
$ git checkout -q upstream/3.10
$ ./configure --help | grep tcl
  --with-tcltk-includes='-I...'
  --with-tcltk-libs='-L...'
```

Reply:
@izumiberat, can you please use `<details>` in your posts? The verbosity and length of your comments make it hard to follow the conversation.

<details>
<summary>Here's a short guide</summary>

```
<details>
<summary>Tests with Python 3.10</summary>

I did this:

    $ git checkout ...
    $ ./configure ... && make

And this happened:

    < put output of build script here >

</details>
```

</details>

Reply:
@erlend-aasland You were right! I was working on `main`. I just tested again for 3.10 and I am all good. Got it for 3.6 --`with-openssl` option. We are all good then.

Reply:
> @izumiberat 
> when running `python3 -m tkinter`, it displays that I am using Tcl/Tk version 8.6. Looking at setup.py, it looks like, it is indeed installed by default if it doesn't find the options:

What version of `tcl-tk` are you installing via Homebrew?

In my case:
```bash
$ brew info tcl-tk | head -n1
tcl-tk: stable 8.6.12 (bottled) [keg-only]
```

This seems to match the version from `python3 -m tkinter`.
When I uninstall it via `brew uninstall tcl-tk` I am no longer able to find the library [1].

So I think everything is OK here.

> @erlend-aasland 
> It may be that setup.py automatically detects and uses the system provided Tcl/Tk libraries on macOS.

In that case, there wouldn't be this issue and PR in the first place, would be?

[1] Yes, `brew  --prefix tcl-tk` still outputs a path but it does regardless the package is installed or not. So, the path doesn't exist.

Reply:
Followed up on my last 👆comment:

I tried to run `configure` with `PKG_CONFIG_PATH` in-line set on 3.7-3.10 and I am getting interesting results when running `PKG_CONFIG_PATH="$(brew --prefix tcl-tk)/lib/pkgconfig" ./configure --with-pydebug --with-openssl=$(brew --prefix openssl@1.1) --with-tcltk-libs="$(pkg-config --libs tcl tk)" --with-tcltk-includes="$(pkg-config --cflags tcl tk)"`:
- 3.7-3.8 - unable to find Tcl/Tk lib, no tkinter module available (expected result)
- 3.9-3.10 - errors with `pkg-config` (as in @erlend-aasland output), however `tkinter` is compiled. When I run `./python -m tkinter` I get a warning: `DEPRECATION WARNING: The system version of Tk is deprecated and may be removed in a future release. Please don't rely on it. Set TK_SILENCE_DEPRECATION=1 to suppress this warning.` and the popup window is a bit broken (black text on black background, only buttons are visible)

Now, when I correctly  configure 3.9/3.10 (`export PKG_CONFIG_PATH`):
- I don't get the deprecation output
- Pop-up window displays correctly with black text on a gray (default) background

After inspecting the compiler output:
```console
gcc -Wno-unused-result -Wsign-compare -g -O0 -Wall -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration -fvisibility=hidden -I./Include/internal -DWITH_APPINIT=1 -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Tcl.framework/Headers -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Tk.framework/Headers -I./Include -I. -I/Users/bxsx/+dev/opensource/cpython/Include -I/Users/bxsx/+dev/opensource/cpython -c /Users/bxsx/+dev/opensource/cpython/Modules/_tkinter.c -o build/temp.macosx-12.4-arm64-3.10-pydebug/Users/bxsx/+dev/opensource/cpython/Modules/_tkinter.o -F /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks -Wno-strict-prototypes
```

@izumiberat I can confirm that in such a situation the default macOS `tkinter` lib is being used.